### PR TITLE
Blocks rendering fix

### PR DESCRIPTION
**Pull request by abidlabs**
* Fixed Blocks so that it can be used inside other blocks as plug-and-play components (Closes: #960)

* Added a demo to illustrate usage: `demo\blocks_plug\run.py`

* Also adds `article` parameter back into `Interface` (Closes: #1093)

### CODE REVIEW

File: demo/blocks_plug/run.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import gradio as gr
+
+text_identity, text_input, text_output = gr.Blocks(), gr.Blocks(), gr.Blocks()
```
Comment (FarukOzderim) [inline demo/blocks_plug/run.py:3]: I think we can rename inner demos better to clarify the message.
```suggestion
identity_demo, input_demo, output_demo= gr.Blocks(), gr.Blocks(), gr.Blocks()
```

File: gradio/blocks.py [hunk 2]
```diff
@@ -19,24 +19,43 @@
...
         self.events = []
 
+    def unrender(self):
+        """
+        Removes self from BlockContext if it has been rendered (otherwise does nothing).
+        Only deletes the first occurrence of self in BlockContext. Removes from the
+        layout and collection of Blocks, but does not delete any event triggers.
+        """
+        self._id = Context.id
+        Context.id += 1
```
Comment (FarukOzderim) [inline gradio/blocks.py:46]: Why increase the id within removal?

**Review by FarukOzderim — COMMENTED** [review]
LGTM, would be better if @aliabid94 takes a look as well.

File: gradio/blocks.py [hunk 3]
```diff
@@ -19,24 +19,43 @@
...
         """
-        self._id = Context.id
-        Context.id += 1
         if Context.block is not None:
             Context.block.children.append(self)
         if Context.root_block is not None:
             Context.root_block.blocks[self._id] = self
         self.events = []
 
+    def unrender(self):
```
[the remark this answers is not in the document]
Comment (FarukOzderim) [reply gradio/blocks.py:39]: What would happen if a component is used in 2 different interfaces.

textbox = gr.Textbox()
with blocks_1:
  demo_1 = Interface(inputs = textbox)
  demo_2 = Interface(inputs = textbox)

**Comment by aliabid94** [comment]
Discussed with @abidlabs earlier, but instead of unrender method, get_component_shortcuts should just create Components with render=False in cunstructor. Should remove the need for unrender. 
LGTM otherwise.

File: demo/blocks_plug/run.py (hunk 1 above)
Comment (abidlabs) [reply demo/blocks_plug/run.py:3]: Sounds good, will do

File: gradio/blocks.py (hunk 2 above)
Comment (abidlabs) [reply gradio/blocks.py:47]: My bad

File: gradio/interface.py [hunk 4]
```diff
@@ -198,8 +198,8 @@ def __init__(
             inputs[inputs.index("state")] = state_variable
             outputs[outputs.index("state")] = state_variable
 
-        self.input_components = [get_component_instance(i) for i in inputs]
-        self.output_components = [get_component_instance(o) for o in outputs]
+        self.input_components = [get_component_instance(i).unrender() for i in inputs]
```
[the remark this answers is not in the document]
Comment (abidlabs) [reply gradio/interface.py:184]: Not sure I follow what you mean

File: gradio/blocks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (abidlabs) [reply gradio/blocks.py:40]: I tested the following code:

```py
textbox = gr.Textbox()

with gr.Blocks() as blocks:
    demo_1 = gr.Interface(lambda x:x, inputs=textbox, outputs="text")
    demo_2 = gr.Interface(lambda x:x, inputs=textbox, outputs="number")

    blocks.launch()
```

And it works fine (although the input textboxes are linked in an interesting way):  

image

**Comment by abidlabs** [comment]
I discussed some more with @aliabid94 and we will actually keep `unrender()`. We need to handle cases like this:

```py
import gradio as gr

blocks = gr.Blocks()
with blocks:    
    gr.Interface(
      fn=lambda x: x,
      inputs=gr.Image(),
      outputs=gr.Image())
blocks.launch() 
```

File: gradio/interface.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (FarukOzderim) [reply gradio/interface.py:184]: oh I misspelled as render, it will be unrender

[the remark this answers is not in the document]
Comment (FarukOzderim) [reply gradio/interface.py:184]: some of the components inside input/output are just string shortcuts("text"), and we shall not `unrender` them

[the remark this answers is not in the document]
Comment (FarukOzderim) [reply gradio/interface.py:184]: Well now I read the discussions below, this point is not correct at all.

File: gradio/components.py [hunk 5]
```diff
@@ -3100,7 +3099,7 @@ def __init__(
         label (str): component name
```
Comment (FarukOzderim) [inline gradio/components.py:3099]: need to update docstring with the label's removal

File: gradio/components.py [hunk 6]
```diff
@@ -3193,11 +3192,10 @@ def __init__(
         components: List[Component],
         samples: List[List[Any]],
         type: str = "values",
-        label: Optional[str] = None,
         css: Optional[Dict] = None,
         **kwargs,
     ):
```
Comment (FarukOzderim) [inline gradio/components.py:3197]: not related to this PR but, Dataset has no docstring, let's add one?

File: gradio/components.py [hunk 7]
```diff
@@ -3274,11 +3272,10 @@ def __init__(
         self,
         component: Component,
         *,
-        label: Optional[str] = None,
         css: Optional[Dict] = None,
         **kwargs,
     ):
```
Comment (FarukOzderim) [inline gradio/components.py:3277]: not related to this PR but, Interpretation has no docstring, let's add one?

**Pull request merged**